tests/e2e/bin/includes.sh

```diff
@@ -0,0 +1,164 @@
+
+        if [ "${2:-}" = "Y" ]; then
+            prompt="Y/n"
+            default=Y
+        elif [ "${2:-}" = "N" ]; then
+            prompt="y/N"
+            default=N
+        else
+            prompt="y/n"
+            default=
+			timeout=
+        fi
+
+		if [ -z "$timeout" ]; then
+        	# Ask the question (not using "read -p" as it uses stderr not stdout)
+        	echo -en "$1 [$prompt] "
+
+        	# Read the answer (use /dev/tty in case stdin is redirected from somewhere else)
+        	read reply </dev/tty
+		else
+			endtime=$((`date +%s` + $timeout));
+			while [ "$endtime" -ge `date +%s` ]; do
+				timediff=$(($endtime - `date +%s`))
+
+				echo -en "\r$1 [$prompt] (Default $default in ${timediff}s) "
+				read -t 1 reply </dev/tty
+
+				if [ -n "$reply" ]; then
+					break
+				fi
+			done
+		fi
+
+        # Default?
+        if [ -z "$reply" ]; then
+            reply=$default
+        fi
+
+        # Check if the reply is valid
+        case "$reply" in
+            Y*|y*) return 0 ;;
+            N*|n*) return 1 ;;
+        esac
+
+    done
+}
+
+##
+# Download from a remote source.
+#
+# Checks for the existence of curl and wget, then downloads the remote file using the first available option.
+#
+# @param {string} remote  The remote file to download.
+# @param {string} [local] Optional. The local filename to use. If it isn't passed, STDOUT is used.
+#
+# @return {bool} Whether the download succeeded or not.
+##
+download() {
+    if command_exists "curl"; then
+        curl -s -o "${2:--}" "$1"
+    elif command_exists "wget"; then
+		wget -nv -O "${2:--}" "$1"
+    fi
+}
+
+##
+# Add error message formatting to a string, and echo it.
+#
+# @param {string} message The string to add formatting to.
+##
+error_message() {
+	echo -en "\033[31mERROR\033[0m: $1"
+}
+
+##
+# Add warning message formatting to a string, and echo it.
+#
+# @param {string} message The string to add formatting to.
+##
+warning_message() {
+	echo -en "\033[33mWARNING\033[0m: $1"
+}
+
+##
+# Add status message formatting to a string, and echo it.
+#
+# @param {string} message The string to add formatting to.
+##
+status_message() {
+	echo -en "\033[32mSTATUS\033[0m: $1"
+}
+
+##
+# Add formatting to an action string.
+#
+# @param {string} message The string to add formatting to.
+##
+action_format() {
+	echo -en "\033[32m$1\033[0m"
+}
+
+##
+# Check if the command exists as some sort of executable.
+#
+# The executable form of the command could be an alias, function, builtin, executable file or shell keyword.
+#
+# @param {string} command The command to check.
+#
+# @return {bool} Whether the command exists or not.
+##
+command_exists() {
+	type -t "$1" >/dev/null 2>&1
+}
+
+# download and install ngrok
+install_ngrok() {
+	if $(command_exists "ngrok"); then
+			NGROK_CMD="ngrok"
+			return
+	fi
+
+	echo -e $(status_message "Installing ngrok...")
+	curl -s https://bin.equinox.io/c/4VmDzA7iaHb/ngrok-stable-linux-amd64.zip > ngrok.zip
```

Comment:
Note that this installs Linux version. Feels like it would be important to highlight in docs that this happens so that folks don't find it surprising or know to setup Ngrok beforehand if they're on Macos or Windows.

Reply:
added a conditional to install ngrok only in CI environment

Reply:
Since it's now only automatically installing on Travis, this probably belongs in `.travis.yml`. Which kind of turns `install_ngrok()` into `check_for_ngrok()`.

Reply:
Travis doesn't use docker, so any of these commands will be invoked locally.

I removed ngrok installation and replaced with the `check_for_ngrok` as suggested in cf10a83